### ISSUE TITLE
Scene2d Listeners via Lambda

### ISSUE THREAD

**Issue by cubesdev** [libgdx/libgdx#6483]
#### Issue details
Currently LibGDX listeners in scene2d are always added via the `addListener(...)` method which takes a generic EventListener interface object. If you want to listen for a specific event you always have to implement an abstract listener class like `ChangeListener`, `ClickListener`, etc...

This creates a lot of boilerpalte code like 

`startGameButton.addListener( new ClickListener() {
    @Override
    public void clicked(InputEvent event, float x, float y) {
         // code comes here
    }
} );`

even for simple click listeners. It is not possible to use lambda methods/method references here which is really annoying (a bit like Java Swing was before Lambda methods were introduced in Java 8). If there were specific addClickListener( ButtonClickListener ) methods then the above code could be written in a single line (ignoring x, y) using lambda notation:

`startGameButton.addClickListener( () -> ... );`

This could easily be implemented on top of the current implementation via simple wrapper classes/methods like:

`public interface ButtonClickListener { ... }`

and in button classes:

`public void addClickListener( final ButtonClickListener listener ) { 
    addListener( new ClickListener() {
        @Override
        public void clicked(InputEvent event, float x, float y) {
            listener.onClick();
        }
    }
);`
 `

 Alternatively one could also migrate ClickListener to an interface with default methods. But since you try to keep Java 8 of LibGDX core the first way is probably better for now.

#### Please select the affected platforms
- [X ] Android
- [X] iOS
- [X] HTML/GWT
- [X] Windows
- [X] Linux
- [X] MacOS

**Comment by NathanSweet** [comment]
I use this in my app code:
```
static public void onChange (Actor actor, Runnable listener) {
	actor.addListener(new ChangeListener() {
		public void changed (ChangeEvent event, Actor actor) {
			listener.run();
		}
	});
}

static public void onClick (Actor actor, Runnable listener) {
	onClick(actor, Buttons.LEFT, listener);
}

static public void onClick (Actor actor, int button, Runnable listener) {
	actor.addListener(new ClickListener(button) {
		public void clicked (InputEvent event, float x, float y) {
			listener.run();
		}
	});
}
```
With a static import it allows eg:
```
onChange(someCheckbox, () -> {
	...
});
```

It doesn't feel great to have many `addXxxListener` methods solely for lambdas. Would `addClickListener` go on Actor? What about `addChangeListener`? I'm not sure Actor should really know about ChangeListener. Likely there are other listener types that definitely don't belong on Actor, in which case we are sprinkling `addXxxListener` methods all over. It's not the worst thing in the world, I'm just not sure we should go down that road. The static import solution above is almost as good, with the mess centralized (in app code, currently).

**Comment by cubesdev** [comment]
@NathanSweet thank you for your feedback and good ideas. 

Sure, addChangeListener etc. should not be added to the base Actor class but rather to the classes which allow change listeners to be added. I understand that you worry that this may introduce some extra code at multiple places. But i think this will actually lead to a more user friendly API since programmers can easily see what listeners are available for the given Actor type (when using an IDE like IntelliJ you get a list of proposed methods which is very useful). For example it's hard to know whether you need a ChangeListener or a ClickListener when using CheckBoxes.

The extra code could also be reduced via Java 8 default methods and interfaces like IsClickable, IsChangeable, or something like that.

**Comment by mgsx-dev** [comment]
sounds familiar to me :D here's my contribution : 
```java
public class UI {
	public static <T extends Actor> T change(T actor, Consumer<ChangeEvent> handler){
		actor.addListener(new ChangeListener() {
			@Override
			public void changed(ChangeEvent event, Actor actor) {
				handler.accept(event);
			}
		});
		return actor;
	}
	public static <T> void select(SelectBox<T> selectBox, Consumer<T> item) {
		change(selectBox, event->item.accept(selectBox.getSelected()));
	}
	public static <T extends Slider> T changeCompleted(T slider, Consumer<ChangeEvent> handler){
		slider.addListener(new ChangeListener() {
			@Override
			public void changed(ChangeEvent event, Actor actor) {
				if(!slider.isDragging()) handler.accept(event);
			}
		});
		return slider;
	}
}
```

**Comment by NathanSweet** [comment]
> For example it's hard to know whether you need a ChangeListener or a ClickListener when using CheckBoxes.

A checkbox has both, so it's not helping with which you need.

I agree it can show you some of the event types that can be fired, but the `addXxxListener` methods won't be an exhaustive list, since arbitrary events can be fired on any actor.

`addXxxListener` methods don't provide benefit to listeners that have multiple methods, like InputListener or ActorGestureListener. I'm even less excited about adding `addInputListener`, `addActorGestureListener`, etc solely for API self documentation.

I would like to be able to pass a lambda directly to an `addListener` method, but I'm not happy about adding `addXxxListener` methods to the generic event listener system when a static import method is nearly as good.

Maybe we could have:
```
actor.addListener(ChangeEvent.class, (event) -> {
	...
});
```
If an event is fired of the specified type, the lambda is called with the event.

**Comment by mgsx-dev** [comment]
as a side note, we recently bumped to Java 7, so lambda is not possible at framework level for now.

**Comment by cubesdev** [comment]
Thank you for your comments!

> as a side note, we recently bumped to Java 7, so lambda is not possible at framework level for now.

The LibGDX language level is not crucial. The proposed changes to make lambda possible for Java 8+ API users can all be implemented in Java 6 as long as interfaces are used instead of the current abstract classes.

> > For example it's hard to know whether you need a `ChangeListener `or a `ClickListener `when using CheckBoxes.
> 
> A checkbox has both, so it's not helping with which you need.

But even this fact is not clear until you know LibGDX internals or read some possibly available docs. Separate addXxxListener methods with some kind of short help text which describe the effect of the listener for the current Actor would help a lot here.

> `addXxxListener` methods don't provide benefit to listeners that have multiple methods, like InputListener or ActorGestureListener. I'm even less excited about adding `addInputListener`, `addActorGestureListener`, etc solely for API self documentation.

They are not really needed since they are a lot less frequently used. In fact restricting Lambda support for simple `ChangeListener `s and `ClickListener`s should already be enough since they are by far the most common listener types (at least in my code).


> Maybe we could have:
> 
> ```
> actor.addListener(ChangeEvent.class, (event) -> {
> 	...
> });
> ```

This looks great and could be a compromise. But the LibGDX event API is inconsistent when it comes to event classes. Not every event type has its own class. There is a `ChangeListener.ChangeEvent` class but no `ClickListener.ClickEvent`. So it won't be possible to distinguish the current available listener types by passing a Class object.

**Comment by NathanSweet** [comment]
> There is a ChangeListener.ChangeEvent class but no ClickListener.ClickEvent.
True, `addListener(ChangeEvent.class, (event) -> {` would only be useful to access an event, not helpful for ClickListener.

It seems like there's not a great solution. One of these seems reasonable:

1) Add methods for only change and click. It doesn't seem great to add them to each scene2d.ui widget. Putting them on Actor isn't great because the listeners are in scene2d.ui. They could go on Widget and WidgetGroup.

2) Provide a class with static methods for use with a static import. This gives us flexibility to have many helpful methods, such as mgsx-dev's `changeCompleted` method above.

Any thoughts?

**Comment by cubesdev** [comment]
> It seems like there's not a great solution. One of these seems reasonable:
> 
> 1. Add methods for only change and click. It doesn't seem great to add them to each scene2d.ui widget. Putting them on Actor isn't great because the listeners are in scene2d.ui. They could go on Widget and WidgetGroup.
> 2. Provide a class with static methods for use with a static import. This gives us flexibility to have many helpful methods, such as mgsx-dev's `changeCompleted` method above.
> 
> Any thoughts?

I recently had another new idea: There could be `ActionListener`s similar like it's done in Java Swing/AWT (documentation: https://docs.oracle.com/javase/tutorial/uiswing/events/actionlistener.html). Most UI elements only have one central event type which could trigger an action event  (clicks for buttons, item change for SelectBox, ...). This action listener could be added via some kind of addActionListener(Runnable) method (maybe it should have a shorter name or overload addListener). Subclasses of Actor can override this method and register their own underlying event listener to fire their appropriate action event. Actors which don't support any of those events at all could simply throw an UnsupportedOperationException or default to ClickListener. What do you think about it?
